lib/eval-modules.nix

```diff
@@ -0,0 +1,59 @@
+callArgs:
+
+let
+  inherit (builtins.foldl' ({ modules, configurers }: m:
+    let
+      match = m: builtins.isAttrs m && m ? configureModules;
+      return = m: {
+        modules = modules ++ [ (removeAttrs m [ "configureModules" ]) ];
+        configurers = configurers ++ [ m.configureModules ];
+      };
+    in if match m then return m else
+    let m' = import m; in if !(builtins.isFunction m || builtins.isAttrs m) && match m'
+    then return (m' // { _file = toString m; key = toString m; })
+    else { modules = modules ++ [ m ]; inherit configurers; }
+  ) { modules = []; configurers = []; } callArgs.modules) modules configurers;
+
+  # Without updating the modules list, the `configureModules` attribute
+  # would not be removed from its module, and it is an invalid option or special attribute,
+  # so it would result in an error.
+  callArgs' = callArgs // { inherit modules; };
+
+  applyConfigureModules = x: if builtins.isFunction x
+    then x callArgs'
+    else callArgs' // x // builtins.listToAttrs (map
+      (name: { inherit name; value = callArgs'.${name} // x.${name}; })
+      (builtins.filter (name: x ? ${name}) [ "args" "specialArgs" ]));
```

Comment:
If `configureModules` is just an attrset rather than a function mapping the existing attrset, I would expect the attributes `args` and `specialArgs` to be merged with the existing attrset.